exchanges/kraken/kraken_websocket.go

```diff
@@ -210,280 +196,257 @@ func isAwaitingCancelOrderResponses(requestID int64, success bool) bool {
 func (k *Kraken) wsHandleData(respRaw []byte) error {
 	if strings.HasPrefix(string(respRaw), "[") {
 		var dataResponse WebsocketDataResponse
-		err := json.Unmarshal(respRaw, &dataResponse)
-		if err != nil {
+		if err := json.Unmarshal(respRaw, &dataResponse); err != nil {
 			return err
 		}
-		if _, ok := dataResponse[0].(float64); ok {
-			err = k.wsReadDataResponse(dataResponse)
-			if err != nil {
-				return err
-			}
+		if len(dataResponse) < 3 {
+			return fmt.Errorf("websocket data array too short: %s", respRaw)
 		}
-		if _, ok := dataResponse[1].(string); ok {
-			err = k.wsHandleAuthDataResponse(dataResponse)
-			if err != nil {
+
+		// For all types of channel second to last field is the channel Name
+		channelName, ok := dataResponse[len(dataResponse)-2].(string)
+		if !ok {
+			return common.GetTypeAssertError("string", dataResponse[len(dataResponse)-2], "channelName")
+		}
+
+		// wsPair is just used for keying the Subs
+		wsPair := currency.EMPTYPAIR
+		if maybePair, ok2 := dataResponse[len(dataResponse)-1].(string); ok2 {
+			var err error
+			if wsPair, err = currency.NewPairFromString(maybePair); err != nil {
 				return err
 			}
 		}
-	} else {
-		var eventResponse map[string]interface{}
-		err := json.Unmarshal(respRaw, &eventResponse)
+
+		c := k.Websocket.GetSubscription(stream.DefaultChannelKey{Channel: channelName, Currency: wsPair, Asset: asset.Spot})
+		if c == nil {
+			return fmt.Errorf("%w: %s %s %s", stream.ErrSubscriptionNotFound, asset.Spot, channelName, wsPair)
+		}
+
+		return k.wsReadDataResponse(c, dataResponse)
+	}
+
+	var eventResponse map[string]interface{}
+	err := json.Unmarshal(respRaw, &eventResponse)
+	if err != nil {
+		return fmt.Errorf("%s - err %s could not parse websocket data: %s",
+			k.Name,
+			err,
+			respRaw)
+	}
+
+	event, ok := eventResponse["event"]
+	if !ok {
+		return nil
```

Comment:
This is the existing behaviour before this PR.
It's not great, but I've hit MaxYakHerdOverflow
This could really do with another pass of abstraction once the dust settles after these changes, mostly to break out the sub-funcs and handle full testing of the parsing and data coming out of DataHandler channel.
I think once these immediate feature fixes are done I can take a look at those tests, or possibly @Beadko